Initialize Solang Aqd

tareknaser opened the pull request:
## Packages

| Package               | Description                                           
| ------------------    | ----------------------------------                    
| `aqd-core`            | The CLI tool core crate                               
| `aqd-polkadot`        | Smart contract interactions for Polkadot              
| `aqd-solana`          | Smart contract interactions for Solana                
| `aqd-utils`           | Utility functions and common code                     
| `aqd-solana-contracts`    | Rust crate for Solana smart contract interactions

tareknaser committed:
initialize the repository

Signed-off-by: Tarek <tareknaser360@gmail.com>

xermicus commented on integration/polkadot/incrementer.contract:
```diff
@@ -0,0 +1 @@
+{"source":{"hash":"0x0cac9f6c156ce7cbc0bbd7ba4f1369aff039eee6d7e55a60855b63962780133d","language":"ink! 5.0.0-alpha","compiler":"rustc 1.73.0","wasm":"0x0061736d01000000011f0660027f7f0060000060047f7f7f7f017f60037f7f7f006000017f60017f00027406057365616c310b6765745f73746f726167650002057365616c301176616c75655f7472616e736665727265640000057365616c3005696e7075740000057365616c320b7365745f73746f7261676500 ...
```
Why ink! contracts and not Solang? Ideally, this repo would just use some examples from Solang and compile them; would provide a more complete test.

tareknaser replied:
I tried to use `solang`-compiled contracts with different versions of `substrate-contracts-node`, but encountered incompatibilities. The errors included:
- With `solang-substrate-ci` ([latest version](https://github.com/hyperledger/solang-substrate-ci/pkgs/container/solang-substrate-ci/128945070?tag=75f66ac)):
```bash
Error uploading the code: Error decoding into dynamic value: Error at : Field deposit_held does not exist in our encoded data
```
- Deploying `solang`-compiled contracts with the latest version of `substrate-contracts-node` failed during instantiation because of differing metadata formats.
```bash
Failed to deserialize ink project metadata from contract metadata
```
- Even when attempting to use an older version of `substrate-contracts-node`, like v26, with `solang`-compiled contracts:
```bash
Error uploading the code: Error decoding into dynamic value: Error at : Field deposit_held does not exist in our encoded data
```
Currently, to test `solang-aqd`, I'm using `ink!`-compiled contracts with the most recent version of `substrate-contracts-node`. I'm using this approach until `solang-substrate-ci` gets updated. 
The goal is to use `solang`-compiled contracts once this update is implemented.
The current setup helps make sure `solang-aqd` works well with Polkadot, handling interactions with Polkadot and providing accurate responses to user requests.

xermicus replied:
> Currently, to test solang-aqd, I'm using ink!-compiled contracts with the most recent version of substrate-contracts-node

> The current setup helps make sure solang-aqd works well with Polkadot, handling interactions with Polkadot and providing accurate responses to user requests.

This is obviously flawed. After all, `solang-aqd` is supposed to be tooling for solang and not `ink!`. The current setup does _not_ make sure that `solang-aqd` works well with Polkadot; it only makes sure that solang-aqd works with some random pre-compiled `ink!` contract. Anyone trying to use `solang-aqd` will just find out that it's broken (despite the CI being green), which is obviously flawed.

I mean testing against an `ink!` contract is fine if your goal is to be compatible with `ink!` contracts too. But it's not the main goal here.

> I'm using this approach until solang-substrate-ci gets updated.

This doesn't have anything to do with the node (if it had, obviously the solution would be to just use a compatible one on the CI). The contracts built by solang are fine and they run fine on all recent versions of pallet contracts.

As you already said. The problem is that `aqd` doesn't understand the metadata format (the node does not know _anything_ about the metadata, the contracts are just binary blobs, they have some requirements but not on the metadata level). The real problem is that `solang-aqd` uses the unreleased version `5` of the metadata format, containing breaking changes; Solang contracts are built with version `4` of `ink_metadata` and do not include the [staticBufferSize](https://docs.rs/ink_metadata/5.0.0-alpha/ink_metadata/struct.EnvironmentSpecBuilder.html#method.static_buffer_size) field in the metadata. Which is fine because this comes from an unreleased version of `ink_metadata` anyways (Solang uses the latest released version which is `4.3.0` and doesn't have this field yet).

If you encounter this problem, it's the wrong approach to just pretend everything is fine. The right way to do this is to state that the tooling does not yet work and to work on fixing this for good. This means you try to upgrade Solang to `ink-5.0.0-alpha` to make it work.

How this CI should be set up is to download a version of solang that is supposed to be compatible with your tool, then compile the contract on the CI and then use that to test everything.

We can merge this PR as it is, however
* This needs to be addressed afterwards
* The README should clearly state that the tool does not yet work for polkadot

seanyoung reviewed (approved):
I've tested it on Solana, and it works very nicely. 

- `aqd solana deploy` does not have the option of saving the generated private key 
- `aqd solana call` does not have a view option (maybe this should be `aqd solana view`)
- There are still clippy warnings in rust 1.73.0

However I feel that these issues are details which can be resolved later. I am happy for this to be merged as is, the code is very nice.

xermicus commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,119 @@
+    branches: [main]
+  # Run every Monday at 1:30 AM UTC
+  schedule:
+    - cron: '30 1 * * 1'
```
I'm not sure if we need this; on PRs and  main should be enough. @seanyoung wdyt?

tareknaser replied:
I removed it for now

xermicus commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,119 @@
+    - name: Check out code
+      uses: actions/checkout@v2
+
+    - name: Set up Rust
```
Adding a step that runs `cargo --version` after installation makes it easier to debug issues

xermicus commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,119 @@
+
+    steps:
+    # Install jq to parse JSON (used in shell scripts)
+    - name: Install jq 
```
This is done for both Polkadot and Solana, could this be done in build instead?

tareknaser replied:
I updated the test workflow to be 3 separate jobs:
- First one checks for formatting, linting and builds for all features
- Second one to test solana node interactions 
	- build the project using `cargo build --no-default-features --features "solana" --release`
- Third one to test polkadot node interactions

I tried it in my github repository. Here is how the [CI](https://github.com/tareknaser/aqd/actions/runs/6825136399) should look like when it's merged

xermicus replied:
LGTM

xermicus commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,119 @@
+
+    - name: Check formatting
+      run: |
+        cargo fmt -- --check
```
This goes for all cargo jobs (fmt, clippy, test): Please add `--workspace`. Also, while currently all feature flags are in the default features, I still think it's better to explicitly have `--all-features` for them too.

Also there should be at least build tests for individual features, e.g. `--no-default-features --features polkadot` and `--no-default-features --features solana`.

xermicus commented on README.md:
```diff
@@ -0,0 +1,18 @@
+It provides a user-friendly interface with commands for deploying smart contracts and calling specific functions on the deployed contracts.
+
+Whether you're developing on Solana or Polkadot, `Aqd` simplifies your smart contract interactions.
+
```
This should have a very short `# Usage` section here with basic installation and usage instructions (can use cargo for that). Otherwise people might just feel unsure about it and not try it out. Also if there is a very simple usage instruction that shows off how to call a contract it might make people think that this is a nice tool they didn't know they needed. But they won't just dream it up so it should be provided.

Additionally, we can add a link once documentation is merged on solang docs

tareknaser replied:
I updated the README file to include `solang` logo, some badges and basic usage.
The installation for now assumes that:
- `aqd` is published to crates.io
- The [documentation PR](https://github.com/hyperledger/solang/pull/1588) that I opened in solang is merged

xermicus replied:
Cool, looks good! Sure, that's fine

xermicus commented on README.md:
```diff
@@ -0,0 +1,18 @@
+# Aqd - Smart Contract CLI Tool
```
Not terribly important - but the logo and some CI badges always look cool in the README :slightly_smiling_face:

xermicus commented on crates/aqd-core/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+
+[dependencies]
+clap = { version = "4.4.7", features = ["derive"] }
+tokio = "1.33.0"
```
This is only used together with the `polkadot` feature, so it should be `optional` and enabled by the `polkadot` feature.

xermicus commented on crates/aqd-utils/src/printing_macros.rs:
```diff
@@ -0,0 +1,44 @@
+
+/// macro to print a title (cyan and bold)
+#[macro_export]
+macro_rules! print_title {
```
I'm not sure all of these macros needed (I guess could be functions instead), but we can leave them as is for now.

xermicus commented on crates/aqd-polkadot/src/commands/call.rs:
```diff
@@ -0,0 +1,150 @@
+    /// If the `execute` flag is set to `false`, it performs a dry run of the call and displays
+    /// the results. If the `output_json` flag is set to `true`, the output is in JSON format.
+    /// Otherwise, it prompts for a transaction confirmation and then submits the transaction for execution.
+    pub async fn handle(&self) -> Result<()> {
```
There are "read only" messages in contracts. They'll have the `mutates` flag set to false in the metadata. For those messages, the `-x/--execute` should lead to an error instead.

xermicus commented on crates/aqd-polkadot/src/commands/mod.rs:
```diff
@@ -0,0 +1,107 @@
+    ///
+    /// If a specific network is chosen, the function returns the URL associated with that network.
+    /// Otherwise, it returns the URL provided by the user in the CLI options.
+    pub fn url(&self) -> Url {
```
nice

xermicus commented on crates/aqd-polkadot/src/commands/mod.rs:
```diff
@@ -0,0 +1,107 @@
+    /// If a specific network is chosen, the function returns the URL associated with that network.
+    /// Otherwise, it returns the URL provided by the user in the CLI options.
+    pub fn url(&self) -> Url {
+        if let Some(net) = &self.network {
+            match net {
+                Network::Rococo => Url::parse("wss://rococo-contracts-rpc.polkadot.io").unwrap(),
+                Network::PhalaPoC5 => Url::parse("wss://poc5.phala.network/ws").unwrap(),
+                Network::AstarShiden => Url::parse("wss://rpc.shiden.astar.network").unwrap(),
+                Network::AstarShibuya => Url::parse("wss://rpc.shibuya.astar.network").unwrap(),
+                Network::Astar => Url::parse("wss://rpc.astar.network").unwrap(),
+                Network::AlephZeroTestnet => Url::parse("wss://ws.test.azero.dev").unwrap(),
+                Network::AlephZero => Url::parse("wss://ws.azero.dev").unwrap(),
+                Network::T3RNT0RN => Url::parse("wss://ws.t0rn.io").unwrap(),
+                Network::PendulumTestnet => {
+                    Url::parse("wss://rpc-foucoco.pendulumchain.tech").unwrap()
+                }
+            }
+        } else {
+            self.url.clone()
+        }
```
You can make this less verbose by letting the `match` just return the URL string and parse it afterwards. Or, even better, implement it in a `fn to_url(&self)` on `Network`

```suggestion
        if let Some(net) = &self.network {
            return Url::parse(match net {
                Network::Rococo => "wss://rococo-contracts-rpc.polkadot.io",
                Network::PhalaPoC5 => "wss://poc5.phala.network/ws",
                Network::AstarShiden => "wss://rpc.shiden.astar.network",
                Network::AstarShibuya => "wss://rpc.shibuya.astar.network",
                Network::Astar => "wss://rpc.astar.network",
                Network::AlephZeroTestnet => "wss://ws.test.azero.dev",
                Network::AlephZero => "wss://ws.azero.dev",
                Network::T3RNT0RN => "wss://ws.t0rn.io",
                Network::PendulumTestnet => "wss://rpc-foucoco.pendulumchain.tech",
            })
            .unwrap();
        }

        self.url.clone()
```

xermicus reviewed (commented):
Looks good!

I'd just like to see some problems around CI and repo setup fixed, also there's a problem with non-mutating messages in Polkadot. The problem is that on the contracts pallet (runtime) level this isn't enforced, so users have to trust the client to not submit an extrinsic for messages considered immutable.

tareknaser committed:
refactor: update Readme, CI Config, and code improvements

This commit includes these changes:

- The README has been updated to include a basic usage guide.
- CI configuration has been optimized to build specifically for Solana and Polkadot features in their respective jobs.
- Restructured the dependencies, making Tokio optional and part of the Polkadot feature.
- Simplified the match statement for URL parsing in the Polkadot feature.
- Addressed and resolved Clippy warnings for improved code quality and consistency.

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser committed:
docs: center the logo and badges in README

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser commented:
I'm aware of some areas for improvement and development in the project. I've got a list of these issues that I'll open on GitHub after the initial PR is merged to keep track of them. Here's what needs addressing:
- Using `contract-extrinsics` Dependency from crates.io Instead of Git
- `aqd` should use `solang` compiled contracts with `solang-substrate-ci` in the CI (`solang-substrate-ci` needs to updated first)
- Remove crate patches in the `Cargo.toml` file once Solana resolves these issues.
- `aqd-polkadot` needs to print a warning if the node version it's trying to communicate with isn't compatible with the CLI version.
- Write a function to do `solana program close` in rust similar to `solana program deploy`
- The function `decode_instruction_return_data` in `crates/aqd-solana-contracts/src/printing_utils.rs` currently retrieves the return data from the logs due to a known bug in the Solana SDK, which is expected to be fixed in [#33639](https://github.com/solana-labs/solana/pull/33639). Once this fix is officially released, the function should be updated to fetch the return data directly from the transaction itself.
-  The file `crates/aqd-utils/src/borsh_encoding.rs` has been copied from `solang`. While `solang` provides the `discriminator` function, it doesn't expose `BorshToken`. Ideally, to reduce its reliance on `solang` (a heavy dependency), `aqd` should aim to be a lightweight crate. Instead, `solang` can depend on `aqd-utils`.
- When displaying information for all possible instructions in Solana, the volume of data might make it challenging to read directly from the CLI. A solution could involve implementing tables to present each instruction's information. One potential crate suitable for this purpose is `radicle-term`.
- For `SolanaTransaction` struct, there should also be a method to dry run the transaction (using `rpc_client.simulate_transaction` function).
- Use Logging
- Doc comments should include simple examples
- `aqd solana call` does not have a view option (maybe this should be `aqd solana view`)
- `aqd solana deploy` does not have the option of saving the generated private key
- There's a problem with non-mutating messages in Polkadot. The problem is that on the contracts pallet (runtime) level this isn't enforced, so users have to trust the client to not submit an extrinsic for messages considered immutable.

xermicus commented:
@tareknaser great, thanks for compiling this exhaustive list. Can you please open them as issues here so we don't loose track?

 > aqd should use solang compiled contracts with solang-substrate-ci in the CI (solang-substrate-ci needs to updated first)

This means upgrading Solang to use (yet unreleased) v5 of `ink_metadata`. Which unfortunately won't work just like that because [this PR](https://github.com/paritytech/ink/commit/6e87ad042ee9ca9bd4e2e7e2632410ea6b62ab0a) broke our use case. You can try to fix that in `ink!` v5 or I can add it to my backlog.

> When displaying information for all possible instructions in Solana, the volume of data might make it challenging to read directly from the CLI. A solution could involve implementing tables to present each instruction's information. One potential crate suitable for this purpose is radicle-term.

Maybe could introduce verbosity level :) Also, sounds like it's related to the tracing functionality you have in mind?

xermicus commented:
@tareknaser thanks, a lot of good stuff in here!
